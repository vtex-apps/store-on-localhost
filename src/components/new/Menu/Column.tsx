import React, { FunctionComponent } from 'react'

const Column: FunctionComponent<Props> = props => {
  return <div>{props.children}</div>
}

interface Props {
  children: React.ReactChildren
}

export default Column
