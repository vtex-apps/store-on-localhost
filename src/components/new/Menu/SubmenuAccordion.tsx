import classNames from 'classnames'
import React from 'react'
import { defineMessages } from 'react-intl'
import Collapsible from './components/Collapsible'

import useCssHandles from '../CssHandles/useCssHandles'

const CSS_HANDLES = ['submenuAccordion'] as const

interface Props {
  isOpen: boolean
  blockClass?: string
}

const SubmenuAccordion: StorefrontFunctionComponent<Props> = ({
  isOpen,
  children,
}) => {
  const handles = useCssHandles(CSS_HANDLES)

  return (
    <Collapsible open={isOpen} >
      <section className={classNames(handles.submenuAccordion, 'w-100 flex pl4 flex')}>
        {children}
      </section>
    </Collapsible>
  )
}

const messages = defineMessages({
  submenuTitle: {
    defaultMessage: '',
    id: 'editor.menu.submenu.title',
  },
})

SubmenuAccordion.getSchema = () => ({
  title: messages.submenuTitle.id,
  type: 'object',
})

export default SubmenuAccordion
