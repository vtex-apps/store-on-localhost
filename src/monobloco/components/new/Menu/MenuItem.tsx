import { path } from 'ramda'
import React, { Reducer, useReducer, useContext } from 'react'

import classNames from 'classnames'

import { defineMessages } from 'react-intl'
// import { ExtensionPoint } from 'vtex.render-runtime'
import { CategoryItemSchema } from './components/CategoryItem'
import { CustomItemSchema } from './components/CustomItem'
import Item from './components/Item'
import { IconProps} from './components/StyledLink'
import useSubmenuImplementation from './hooks/useSubmenuImplementation'

import useCssHandles from '../CssHandles/useCssHandles'
import MenuContext from './components/MenuContext'

const CSS_HANDLES = ['menuItem', 'menuItemInnerDiv']

const submenuInitialState = {
  hasBeenActive: false,
  isActive: false,
}

type SubmenuState = typeof submenuInitialState

type SubmenuAction = 
  | {type: 'SHOW_SUBMENU'}
  | {type: 'HIDE_SUBMENU'}

const submenuReducer: Reducer<SubmenuState, SubmenuAction> =  (state, action) => {
  switch (action.type) {
    case 'SHOW_SUBMENU':
      return {
        hasBeenActive: true,
        isActive: true,
      }
    case 'HIDE_SUBMENU':
      return {
        ...state,
        isActive: false,
      }
    default:
      return state
  }
}

const MenuItem: StorefrontFunctionComponent<MenuItemSchema> = ({
  children,
  ...props
}) => {
  const { experimentalOptimizeRendering } = useContext(MenuContext)
  const [{ isActive, hasBeenActive }, dispatch] = useReducer(submenuReducer, submenuInitialState)
  const handles = useCssHandles(CSS_HANDLES)

  const setActive = (value: boolean) => {
    dispatch({ type: value ? 'SHOW_SUBMENU' : 'HIDE_SUBMENU' })
  }

  /* This is a temporary check of which kind of submenu is being
   * inserted. This will be replaced by new functionality of useChildBlocks
   * in the future. */
  const submenuImplementation = useSubmenuImplementation()
  const isCollapsible = submenuImplementation === 'submenu.accordion'
  console.log('teste CHILDREN: ', children)

  if (isCollapsible) {
    console.log('teste isCollapsible: ')
    return (
      <li className={classNames(handles.menuItem, 'list')}>
        <div className={handles.menuItemInnerDiv}
          onClick={event => {
            setActive(!isActive)
            event.stopPropagation()
          }}>
          <Item {...props} accordion active={isActive} />
        </div>
        {(hasBeenActive || !experimentalOptimizeRendering) && children && ( /* Collapsible menus need to still persist after being open,
                             * to make the closing transition work properly */
          <>
            {/* <ExtensionPoint id="submenu" isOpen={isActive} />
            <ExtensionPoint id="unstable--submenu" isOpen={isActive} /> */}
            {/* <S */}
            {/* {props.submenuProps ? <Submenu {...props.submenuProps} /> : null} */}
            {/* {children} */}
            {React.cloneElement(children, { isOpen: isActive})}
          </>
        )}
      </li>
    )
  }

  return (
    <li
      className={classNames(handles.menuItem, 'list')}
      onMouseEnter={() => setActive(true)}
      onMouseLeave={() => setActive(false)}>
      <Item {...props} active={isActive} />
      {(isActive || !experimentalOptimizeRendering) && children && (
        <>
        {/* {children} */}
          {React.cloneElement(children, { isOpen: isActive})}
          {/* <ExtensionPoint id="submenu" isOpen={isActive} />
          <ExtensionPoint id="unstable--submenu" isOpen={isActive} /> */}
        </>
      )}
    </li>
  )
}

export interface MenuItemSchema {
  id: string
  type: 'category' | 'custom'
  iconProps: IconProps
  iconPosition: 'left' | 'right'
  highlight: boolean
  itemProps: CategoryItemSchema | CustomItemSchema
  blockClass?: string
  experimentalOptimizeRendering?: boolean
  submenuProps?: any
}

const messages = defineMessages({
  categoryIdTitle: {
    defaultMessage: '',
    id: 'admin/editor.menu.item.params.categoryId.title',
  },
  categoryTitle: {
    defaultMessage: '',
    id: 'admin/editor.menu.item.category.title',
  },
  customTitle: {
    defaultMessage: '',
    id: 'admin/editor.menu.item.custom.title',
  },
  customTypeTitle: {
    defaultMessage: '',
    id: 'admin/editor.menu.item.params.type.title',
  },
  externalTitle: {
    defaultMessage: '',
    id: 'admin/editor.menu.item.params.external.title',
  },
  highlightTitle: {
    defaultMessage: '',
    id: 'admin/editor.menu.item.highlight.title',
  },
  hrefTitle: {
    defaultMessage: '',
    id: 'admin/editor.menu.item.params.href.title',
  },
  iconIdTitle: {
    defaultMessage: '',
    id: 'admin/editor.menu.item.iconId.title',
  },
  internalTitle: {
    defaultMessage: '',
    id: 'admin/editor.menu.item.params.internal.title',
  },
  itemIdTitle: {
    defaultMessage: '',
    id: 'admin/editor.menu.item.id.title',
  },
  itemsTitle: {
    defaultMessage: '',
    id: 'admin/editor.menu.items.title',
  },
  noFollowTitle: {
    defaultMessage: '',
    id: 'admin/editor.menu.item.params.noFollow.title',
  },
  paramsTextTitle: {
    defaultMessage: '',
    id: 'admin/editor.menu.item.params.text.title',
  },
  paramsTitle: {
    defaultMessage: '',
    id: 'admin/editor.menu.item.params.title',
  },
  tagTitleTitle: {
    defaultMessage: '',
    id: 'admin/editor.menu.item.params.tagTitle.title',
  },
  typeTitle: {
    defaultMessage: '',
    id: 'admin/editor.menu.item.type.title',
  },
})

MenuItem.getSchema = props => {
  const text = path(['itemProps', 'text'], props)
  const id = props && props.id ? props.id : ''

  // tslint:disable: object-literal-sort-keys
  return {
    title: [text, id, messages.itemsTitle.id].find(e => !!e),
    type: 'object',
    required: ['type'],
    properties: {
      id: {
        default: id,
        title: messages.itemIdTitle.id,
        type: 'string',
      },
    },
  }
}

export default MenuItem

