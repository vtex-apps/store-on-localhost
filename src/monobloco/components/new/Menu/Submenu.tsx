import classNames from 'classnames'
import React from 'react'
import { defineMessages } from 'react-intl'

import useCssHandles from '../CssHandles/useCssHandles'

const CSS_HANDLES = ['submenu', 'submenuWrapper', 'submenuContainer'] as const

const MAX_TACHYONS_SCALE = 11

enum Orientation {
  horizontal = 'horizontal',
  vertical = 'vertical',
}

export type TachyonsScaleInput = string | number | undefined

/** TODO: This function is copied from the flex-layout app.
 * It should probably be exported to a separate package.
 */
const parseTachyonsValue = (value: TachyonsScaleInput, name?: string) => {
  if (!value) {
    return 0
  }

  const parsedValue = typeof value === 'string' ? parseInt(value, 10) : value

  if (isNaN(parsedValue) || String(parsedValue) !== String(value) || parsedValue < 0 || parsedValue > MAX_TACHYONS_SCALE) {
    if (name) {
      console.warn(`Invalid ${name} value. It should be an integer between 0 and ${MAX_TACHYONS_SCALE}.`)
    }
    return 0
  }

  return parsedValue
}


const Submenu: StorefrontFunctionComponent<SubmenuProps> = ({
  isOpen,
  width,
  children,
  orientation = Orientation.horizontal,
  paddingTop = 4,
  paddingBottom = 4,
}) => {
  console.log('teste SHOWING SUBMENU')
  const handles = useCssHandles(CSS_HANDLES)

  return (
    <div className={`${handles.submenuContainer} ${width === '100%' ? '' : 'relative'}`}>
      <div
        className={classNames(handles.submenuWrapper, `absolute left-0 bg-base pt${parseTachyonsValue(paddingTop, 'paddingTop')} pb${parseTachyonsValue(paddingBottom, 'paddingBottom')} bw1 bb b--muted-3 z-2`,
          {
            dn: !isOpen,
            flex: isOpen,
            'w-100': width === '100%',
            'w-auto ml6': width === 'auto',
          }
        )}
      >
        <section className={classNames(handles.submenu, 'w-100 flex justify-center', { 'flex-column': orientation === Orientation.vertical })}>{children}</section>
      </div>
    </div>
  )
}

export interface SubmenuProps extends SubmenuSchema {
  isOpen: boolean
  orientation: Orientation
  paddingTop: number | string
  paddingBottom: number | string
  blockClass?: string
}

interface SubmenuSchema {
  width: '100%' | 'auto'
}

const messages = defineMessages({
  submenuTitle: {
    defaultMessage: '',
    id: 'admin/editor.menu.submenu.title',
  },
  submenuWidthTitle: {
    defaultMessage: '',
    id: 'admin/editor.menu.item.submenuWidth.title',
  },
})

// tslint:disable: object-literal-sort-keys
Submenu.getSchema = () => ({
  title: messages.submenuTitle.id,
})

export default Submenu
