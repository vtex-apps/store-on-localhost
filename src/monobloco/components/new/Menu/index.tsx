import PropTypes from 'prop-types'
import React, { ReactNode } from 'react'

import useCssHandles from '../CssHandles/useCssHandles'
import { Link } from '../../../../vtex.render-runtime'
import Container from '../StoreComponents/Container'
import Options from './constants/Options'

// This is required because is used in static schema attribute of Menu Component
const GLOBAL_PAGES =
  (global as any).__RUNTIME__ && Object.keys((global as any).__RUNTIME__.pages)

const MAX_ITEMS: number = 10

const CSS_HANDLES = [
  'container',
  'linkLeft',
  'linkMiddle',
  'linkRight',
  'menuLinkNav',
  'menuLinkDivLeft',
  'menuLinkDivMiddle',
  'menuLinkDivRight',
  'renderLink',
] as const

interface Link {
  text?: string | null
  internalPage: string | null
  params?: string | null
  externalPage?: string | null
  typeOfRoute?: string | null
  page?: string | null
  position?: string | null
}

interface Options {
  LEFT: string
  MIDDLE: string
  RIGHT: string
  INTERNAL: string
  EXTERNAL: string
}

interface Props {
  links?: Link[]
}

/**
 * Convert the string params to a js object
 */
const getParams = (params?: string | null): { [key: string]: string } => {
  const json: { [key: string]: string } = {}
  if (params) {
    const array: string[] = params.split(',')
    array.forEach((item: string) => {
      const pair: string[] = item.split('=')
      json[pair[0]] = pair[1]
    })
  }
  return json
}

const getValidPage = (page?: string | null): string => {
  if (!page || (!page.startsWith('http://') && !page.startsWith('https://'))) {
    page = `http://${page}`
  }
  return page
}

/**
 * Links MenuLink Component.
 * Shows a menu bar with links.
 */
const MenuLink: StorefrontFunctionComponent<Props> = ({ links = [] }) => {
  const handles = useCssHandles(CSS_HANDLES)

  const renderLink = (link: Link, index: number): ReactNode => {
    let className: string = `${handles.renderLink} t-small link c-muted-2 dib dim mr3 mr4-ns`
    switch (link.position) {
      case Options.LEFT:
        className = `${handles.linkLeft} ${className}`
        break
      case Options.MIDDLE:
        className = `${handles.linkMiddle} ${className}`
        break
      case Options.RIGHT:
        className = `${handles.linkRight} ${className}`
        break
    }
    return link.typeOfRoute === Options.INTERNAL ? (
      <Link
        className={className}
        key={`${link.text}-${link.position}-${index}`}
        page={link.internalPage}
        params={getParams(link.params)}
      >
        {link.text}
      </Link>
    ) : (
      <a
        className={className}
        key={`${link.text}-${link.position}-${index}`}
        href={getValidPage(link.externalPage)}
        target="_blank"
      >
        {link.text}
      </a>
    )
  }

  if (!links.length) {
    return null
  }
  return (
    <div className={`${handles.container} bg-base h2 c-muted-2 w-100 dn db-ns`}>
      <Container>
        <nav className={`${handles.menuLinkNav} flex justify-between`}>
          <div className={`${handles.menuLinkDivLeft} flex-grow pa3 flex items-center`}>
            {links
              .filter(link => link.position === Options.LEFT)
              .map((link, index) => {
                return renderLink(link, index)
              })}
          </div>
          <div className={`${handles.menuLinkDivMiddle} flex-grow pa3 flex items-center`}>
            {links
              .filter(link => link.position === Options.MIDDLE)
              .map((link, index) => {
                return renderLink(link, index)
              })}
          </div>
          <div className={`${handles.menuLinkDivRight} flex-grow pa3 flex items-center`}>
            {links
              .filter(link => link.position === Options.RIGHT)
              .map((link, index) => {
                return renderLink(link, index)
              })}
          </div>
        </nav>
      </Container>
    </div>
  )
}

// MenuLink.propTypes = {
//   links: PropTypes.arrayOf(
//     PropTypes.shape({
//       /** External page to redirect */
//       externalPage: PropTypes.string,
//       /** Internal page to redirect */
//       internalPage: PropTypes.string.isRequired,
//       /** Page route to redirect when clicked */
//       page: PropTypes.string,
//       /** Params to redirect to internal page */
//       params: PropTypes.string,
//       /** Link position  */
//       position: PropTypes.string,
//       /** Link text */
//       text: PropTypes.string,
//       /** Type of Route (internal or external) */
//       typeOfRoute: PropTypes.string,
//     }).isRequired
//   ),
// }

// tslint:disable: object-literal-sort-keys
MenuLink.schema = {
  title: 'admin/editor.menu-link',
  description: 'admin/editor.menu-link.description',
  type: 'object',
  properties: {
    links: {
      title: 'admin/editor.menu-link.links',
      type: 'array',
      minItems: 0,
      maxItems: MAX_ITEMS,
      items: {
        title: 'admin/editor.menu-link.links.link',
        type: 'object',
        properties: {
          text: {
            title: 'admin/editor.menu-link.links.link.text',
            type: 'string',
          },
          internalPage: {
            title: 'admin/editor.menu-link.links.link.internalPage',
            description:
              'admin/editor.menu-link.links.link.internalPage.description',
            type: 'string',
            enum: GLOBAL_PAGES,
          },
          params: {
            title: 'admin/editor.menu-link.links.link.params',
            description: 'admin/editor.menu-link.links.link.params.description',
            type: 'string',
          },
          externalPage: {
            title: 'admin/editor.menu-link.links.link.externalPage',
            description:
              'admin/editor.menu-link.links.link.externalPage.description',
            type: 'string',
          },
          typeOfRoute: {
            title: 'admin/editor.menu-link.links.link.typeOfRoute',
            type: 'string',
            enum: [Options.INTERNAL, Options.EXTERNAL],
            enumNames: [
              'admin/editor.menu-link.links.link.typeOfRoute.internal',
              'admin/editor.menu-link.links.link.typeOfRoute.external',
            ],
            default: Options.INTERNAL,
            widget: {
              'ui:widget': 'radio',
              'ui:options': {
                inline: true,
              },
            },
          },
          position: {
            title: 'admin/editor.menu-link.links.link.position',
            type: 'string',
            enum: [Options.LEFT, Options.MIDDLE, Options.RIGHT],
            enumNames: [
              'admin/editor.menu-link.links.link.position.left',
              'admin/editor.menu-link.links.link.position.middle',
              'admin/editor.menu-link.links.link.position.right',
            ],
            default: Options.MIDDLE,
          },
        },
      },
    },
  },
}

export default MenuLink
