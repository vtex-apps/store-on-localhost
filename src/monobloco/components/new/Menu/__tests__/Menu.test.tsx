import { render } from '@vtex/test-tools/react'
import React from 'react'

import Menu from '../Menu'

describe('Menu Component', () => {
  function renderComponent() {
    return render(<Menu />)
  }

  it('should be rendered', () => {
    expect(renderComponent()).toBeDefined()
  })

  it('should match the snapshot', () => {
    const { asFragment } = renderComponent()
    expect(asFragment()).toMatchSnapshot()
  })

  it('should be editable through storefront', () => {
    expect(Menu.getSchema({})).toHaveProperty('title')
  })
})
