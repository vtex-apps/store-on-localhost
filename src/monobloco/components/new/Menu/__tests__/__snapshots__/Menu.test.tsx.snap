// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu Component should match the snapshot 1`] = `
<DocumentFragment>
  <nav
    class="menuContainerNav"
  >
    <ul
      class="menuContainer list flex pl0 mv0 flex-row"
    />
  </nav>
</DocumentFragment>
`;
